storyblok/field-plugin
# EXT-929: Plugin Loading and Error States

## What?

Introduces a more explicit way to handle the three states a field plugin can be in:

- loading - the initial state that a plugin first will be in, before the value and other data has been retreived via messaging to the container
- error - if the loading failed. Can also be reserved for other scenarios where the plugin fails during its lifetime.
- loaded - the state where the user can interact with the field plugin.

Previously, if the plugin was still loading or failed to load, the application would simply be stuck in a state where the `state` and `actions` were undefined; without a clear definition of what this meant. 

When implementing a field plugin, you will need to check whether it finished loading and only then render the field plugin itself.

This pull request also simplifies the integrations between js framework and the @storyblok/field-plugin library. To integrate, simply

. Set the initial state to loading
. Run the `createFieldPlugin()` effect
. In the createFieldPlugin callback, update the state.

## Why?

- Makes it easier to handle loading and error states when implementing a field plugin.
- Makes it easier to integrate with frameworks.

## How to test? (optional)

For usage, see `packages/demo/src/useFieldPlugin.ts`

The React `demo` application in `packages/` should work as before.


```shell
yarn
yarn build && yarn build
yarn test
yarn lint
yarn check:types
yarn dev:container
yarn dev:demo
```

To see the error state, open the demo in a standalone window - i.e. without being embedded within the container. http://localhost:8080/
